### PR TITLE
Improve Dockerfile and PyCharm experience

### PR DESCRIPTION
**Pull request by lsorber**
1. Splits the Dockerfile into `base`, `dev`, `ci`, and `app` stages
2. Installs the development environment as part of the `dev` stage – this improves the PyCharm experience
3. Automatically copies back the generated `poetry.lock` file from (2) back to the host when you start the container so that it can be committed
4. Should also be a step in the direction of Windows compatibility

### CODE REVIEW

**Review by tanguyNaets — APPROVED** [review]
LGTM, but I'm not a pycharm user, so I can't comment on the potential improved experience for them.

**Review by sinopeus — APPROVED** [review]

File: {{ cookiecutter.package_name|slugify }}/docker-compose.yml [hunk 1]
```diff
@@ -12,7 +12,12 @@ services:
         {%- endif %}
       context: .
       target: dev
-    command: [ "sleep", "infinity" ]
+    command:
+      [
+        "sh",
+        "-c",
+        "cp --update /var/lib/poetry/poetry.lock && cp --update /var/lib/git/* /app/.git/hooks/ && sleep infinity"
```
Comment (sinopeus) [inline {{ cookiecutter.package_name|slugify }}/docker-compose.yml:19]: I think you're missing the destination for the first `cp` here.

**Review by sinopeus — CHANGES_REQUESTED** [review]

File: {{ cookiecutter.package_name|slugify }}/Dockerfile [hunk 2]
```diff
@@ -48,12 +42,37 @@ ARG POETRY_PYPI_TOKEN_PYPI
...
+    --mount=type=secret,id=poetry-http-basic-username \
+    --mount=type=secret,id=poetry-http-basic-password \
+    {%- endif %}
+    mkdir -p src/{{ cookiecutter.package_name|slugify|replace("-", "_") }}/ && touch src/{{ cookiecutter.package_name|slugify|replace("-", "_") }}/__init__.py && touch README.md && \
+    {%- if cookiecutter.private_package_repository_name %}
+    POETRY_HTTP_BASIC_{{ cookiecutter.private_package_repository_name|upper }}_USERNAME=${POETRY_HTTP_BASIC_{{ cookiecutter.private_package_repository_name|upper }}_USERNAME:-$(cat /run/secrets/poetry-http-basic-username)} \
+    POETRY_HTTP_BASIC_{{ cookiecutter.private_package_repository_name|upper }}_PASSWORD=${POETRY_HTTP_BASIC_{{ cookiecutter.private_package_repository_name|upper }}_PASSWORD:-$(cat /run/secrets/poetry-http-basic-password)} \
+    {%- endif %}
+    poetry install --no-interaction && \
+    pre-commit install --install-hooks --hook-type commit-msg --hook-type pre-commit && \
```
[the remark this answers is not in the document]
Comment (sinopeus) [reply {{ cookiecutter.package_name|slugify }}/Dockerfile]: I'd leave it like this - this should work fine enough (will test).

File: {{ cookiecutter.package_name|slugify }}/docker-compose.yml (hunk 1 above)
Comment (lsorber) [reply {{ cookiecutter.package_name|slugify }}/docker-compose.yml:19]: Oh, good catch, got lost in my edit.

Comment (lsorber) [reply {{ cookiecutter.package_name|slugify }}/docker-compose.yml]: Fixed!

File: {{ cookiecutter.package_name|slugify }}/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (sinopeus) [reply {{ cookiecutter.package_name|slugify }}/Dockerfile]: Tested and approved!

**Review by sinopeus — APPROVED** [review]

**Pull request merged**